File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/consumer/types/TypeBindingSpec.java
package io.micronaut.rabbitmq.docs.consumer.types;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;

@MicronautTest
@Property(name = "spec.name", value = "TypeBindingSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class TypeBindingSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testBindingByType(ProductClient productClient, ProductListener productListener) {

// tag::producer[]
        productClient.send("body".getBytes(), "text/html");
        productClient.send("body2".getBytes(), "application/json");
        productClient.send("body3".getBytes(), "text/xml");
// end::producer[]


        await().atMost(60, SECONDS).until(() ->
                productListener.messages.size() == 3 &&
                productListener.messages.contains("exchange: [], routingKey: [product], contentType: [text/html]") &&
                productListener.messages.contains("exchange: [], routingKey: [product], contentType: [application/json]") &&
                productListener.messages.contains("exchange: [], routingKey: [product], contentType: [text/xml]")
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/consumer/types/TypeBindingSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="TypeBindingSpec")
class TypeBindingSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_binding_by_type(self) -> None:
# tag::producer[]
        self.product_client.send(b"body", "text/html")
        self.product_client.send(b"body2", "application/json")
        self.product_client.send(b"body3", "text/xml")
# end::producer[]

        messages = self.product_listener.messages
        Await.until(lambda: len(messages) == 3
                    and "exchange: [], routingKey: [product], contentType: [text/html]" in messages
                    and "exchange: [], routingKey: [product], contentType: [application/json]" in messages
                    and "exchange: [], routingKey: [product], contentType: [text/xml]" in messages)
